ocaml/xapi/livepatch.ml

```diff
@@ -0,0 +1,343 @@
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "livepatch" end)
+
+open D
+
+exception Invalid_component
+
+type component = Xen | Kernel
+
+let component_of_string = function
+  | "Xen" | "xen" ->
+      Xen
+  | "Kernel" | "kernel" ->
+      Kernel
+  | _ ->
+      raise Invalid_component
+
+let pkg_name_of_component = function
+  | Xen ->
+      "xen-hypervisor"
+  | Kernel ->
+      "kernel"
+
+let string_of_component = function Xen -> "xen" | Kernel -> "kernel"
+
+type t = {
+    component: component
+  ; base_build_id: string
+  ; base_version: string
+  ; base_release: string
+  ; to_version: string option
+  ; to_release: string option
+}
+
+let to_json lp =
+  let option_to_json o =
+    Option.(value (map (fun x -> `String x) o) ~default:`Null)
+  in
+  `Assoc
+    [
+      ("component", `String (string_of_component lp.component))
+    ; ("base_build_id", `String lp.base_build_id)
+    ; ("base_version", `String lp.base_version)
+    ; ("base_release", `String lp.base_release)
+    ; ("to_version", option_to_json lp.to_version)
+    ; ("to_release", option_to_json lp.to_release)
+    ]
+
+let of_json js =
+  try
+    let open Yojson.Basic.Util in
+    {
+      component= member "component" js |> to_string |> component_of_string
+    ; base_build_id= member "base_build_id" js |> to_string
+    ; base_version= member "base_version" js |> to_string
+    ; base_release= member "base_release" js |> to_string
+    ; to_version= member "to_version" js |> to_string_option
+    ; to_release= member "to_release" js |> to_string_option
+    }
+  with e ->
+    let msg = "Can't construct a livepatch from json" in
+    error "%s: %s" msg (ExnHelper.string_of_exn e) ;
+    raise Api_errors.(Server_error (internal_error, [msg]))
+
+let get_latest_livepatch lps =
+  List.map (fun (_, _, t_v, t_r) -> (t_v, t_r)) lps
+  |> Rpm.get_latest_version_release
+  |> function
+  | Some (v, r) ->
+      List.find_opt (fun (_, _, t_v, t_r) -> t_v = v && t_r = r) lps
+  | None ->
+      None
+
+module KernelLivePatch = struct
+  let get_running_livepatch () =
+    let r =
+      (* This expects the kernel livepatch module name should be built as:
+       * livepatch_<base_version>__<base_release>__<to_version>__<to_releaes>
+       * The "." in <base_version>, <base_release>, <to_version>, and <to_releaes>
+       * should be replaced with "_".
+       * For example:
+       * livepatch_4_19_19__8_0_20__4_19_19__8_0_22
+       *)
+      Re.Posix.compile_pat
+        {|^[ ]*livepatch_([^ \[]+)__([^ \[]+)__([^ \[]+)__([^ \[]+).*$|}
```

Comment:
Can we make this regex stronger? `([0-9_]+)` seems to capture the expected version strings.

Reply:
The version and release strings may contain other characters. Actually we don't have any expectations on the strings.  Here we only need to separate them.

Reply:
Another option might be to chop off the `livepatch_` prefix and then split the string on `__`. I suppose it doesn't really matter.

Reply:
If we can describe the operation with a regular expression (because the number of fields  is known in advance) I would prefer that over a procedural approach of chopping an splitting - let the regex engine do the work. A regex doesn't work if this becomes context sensitive - earlier information informs later decisions. In that case I would again consider using Angstrom because it is still quite declarative.